src/main/java/io/deepstream/Record.java

```diff
@@ -509,7 +509,9 @@ Record whenReady(RecordReadyListener recordReadyListener) {
         if( this.isReady ) {
             recordReadyListener.onRecordReady( this.name, this );
         } else {
-            this.onceRecordReadyListeners.add( recordReadyListener );
+            synchronized (this) {
```

Comment:
I'm not 100% sure either, the problem is here.

```
private void onRead( Message message ) {
    ackTimeoutRegistry.clear( message );

    Map<String,JsonElement> oldValues = beginChange();
    this.version = Integer.parseInt( message.data[ 1 ] );
    this.data = gson.fromJson( message.data[ 2 ], JsonElement.class );
    this.path.setCoreElement(this.data);
    completeChange( oldValues );
    setReady();
}

private void setReady() {
    this.isReady = true;

    for(RecordReadyListener recordReadyListener: this.onceRecordReadyListeners) {
        recordReadyListener.onRecordReady( this.name, this );
    }
    this.onceRecordReadyListeners.clear();
}
```

When there are multiple threads with a reference to a record, often the `recordReadyListener` is null. Causing a NPE.

By putting synchronized, this never happens.